cluster-autoscaler/ca.patch

```diff
@@ -0,0 +1,145 @@
+ 	podListWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", namespace, selector)
++	podListWatch = wrapListWatchWithNeonVMUsage(podListWatch)
+ 	store, reflector := cache.NewNamespaceKeyedIndexerAndReflector(podListWatch, &apiv1.Pod{}, time.Hour)
+ 	podLister := v1lister.NewPodLister(store)
+ 	go reflector.Run(stopchannel)
+@@ -209,6 +215,7 @@ func NewScheduledPodLister(kubeClient client.Interface, stopchannel <-chan struc
+ 	selector := fields.ParseSelectorOrDie("spec.nodeName!=" + "" + ",status.phase!=" +
+ 		string(apiv1.PodSucceeded) + ",status.phase!=" + string(apiv1.PodFailed))
+ 	podListWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", apiv1.NamespaceAll, selector)
++	podListWatch = wrapListWatchWithNeonVMUsage(podListWatch)
+ 	store, reflector := cache.NewNamespaceKeyedIndexerAndReflector(podListWatch, &apiv1.Pod{}, time.Hour)
+ 	podLister := v1lister.NewPodLister(store)
+ 	go reflector.Run(stopchannel)
+@@ -218,6 +225,104 @@ func NewScheduledPodLister(kubeClient client.Interface, stopchannel <-chan struc
+ 	}
+ }
+ 
++// copied from github.com/neondatabase/autoscaling, neonvm/apis/neonvm/v1/virtualmachine_types.go.
++//
++// this is duplicated so we're not *also* managing an additional dependency.
++type virtualMachineUsage struct {
++	// NOTE: update this after neondatabase/autoscaling#172 merges
++	CPU    int32             `json:"cpu"`
++	Memory resource.Quantity `json:"memory"`
++}
++
++func wrapListWatchWithNeonVMUsage(lw *cache.ListWatch) *cache.ListWatch {
++	updatePodRequestsFromNeonVMAnnotation := func(pod *apiv1.Pod) {
++		annotation, ok := pod.Annotations["vm.neon.tech/usage"]
++		if !ok {
++			return
++		}
++
++		var usage virtualMachineUsage
++		if err := json.Unmarshal([]byte(annotation), &usage); err != nil {
++			return
++		}
++
++		pod.Spec.Containers[0].Resources.Requests = apiv1.ResourceList(map[apiv1.ResourceName]resource.Quantity{
++			// NOTE: update this after neondatabase/autoscaling#172 merges
++			apiv1.ResourceCPU:    *resource.NewMilliQuantity(int64(usage.CPU)*1000, resource.DecimalSI),
++			apiv1.ResourceMemory: usage.Memory,
++		})
++	}
++
++	return &cache.ListWatch{
++		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
++			obj, err := lw.List(options)
++			if err != nil {
++				return obj, err
++			}
++
++			list := obj.(*apiv1.PodList)
++			for i := range list.Items {
++				updatePodRequestsFromNeonVMAnnotation(&list.Items[i])
++			}
++			return obj, nil
++		},
++		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
++			iface, err := lw.Watch(options)
++			if err != nil {
++				return iface, err
++			}
++
++			// Wrap the channel to update the pods as they come through
++			wrappedEvents := make(chan watch.Event)
++			proxyIface := watch.NewProxyWatcher(wrappedEvents)
++
++			go func() {
++				defer iface.Stop()
++
++				events := iface.ResultChan()
++
++				for {
++					var ev watch.Event
```

Comment:
I would declare on 107 and nest the remainder of the function in that arm of the select statment.